x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentile.java

```diff
@@ -0,0 +1,408 @@
+        var index = p * (valueCount - 1);
+        var lowerIndex = (int) index;
+        var upperIndex = lowerIndex + 1;
+        var fraction = index - lowerIndex;
+
+        if (valuesBlock.mvSortedAscending()) {
+            if (percentile == 0) {
+                builder.appendDouble(valuesBlock.getDouble(0));
+            } else if (percentile == 100) {
+                builder.appendDouble(valuesBlock.getDouble(valueCount - 1));
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                var result = calculateDoublePercentile(fraction, valuesBlock.getDouble(lowerIndex), valuesBlock.getDouble(upperIndex));
+                builder.appendDouble(result);
+            }
+        }
+
+        var values = new double[valueCount];
+
+        for (int i = 0; i < valueCount; i++) {
+            values[i] = valuesBlock.getDouble(firstValueIndex + i);
+        }
+
+        Arrays.sort(values);
+
+        if (percentile == 0) {
+            builder.appendDouble(values[0]);
+        } else if (percentile == 100) {
+            builder.appendDouble(values[valueCount - 1]);
+        } else {
+            assert lowerIndex >= 0 && upperIndex < valueCount;
+            var result = calculateDoublePercentile(fraction, values[lowerIndex], values[upperIndex]);
+            builder.appendDouble(result);
+        }
+    }
+
+    private static void processInts(IntBlock.Builder builder, int position, IntBlock valuesBlock, double percentile) {
+        int valueCount = valuesBlock.getValueCount(position);
+        int firstValueIndex = valuesBlock.getFirstValueIndex(position);
+
+        if (valueCount == 0) {
+            builder.appendNull();
+            return;
+        }
+
+        if (percentile < 0 || percentile > 100) {
+            throw new IllegalArgumentException("Percentile parameter must be a number between 0 and 100, found [" + percentile + "]");
+        }
+
+        if (valueCount == 1) {
+            builder.appendInt(valuesBlock.getInt(firstValueIndex));
+            return;
+        }
+
+        var p = percentile / 100.0;
+        var index = p * (valueCount - 1);
+        var lowerIndex = (int) index;
+        var upperIndex = lowerIndex + 1;
+        var fraction = index - lowerIndex;
+
+        if (valuesBlock.mvSortedAscending()) {
+            if (percentile == 0) {
+                builder.appendInt(valuesBlock.getInt(0));
+            } else if (percentile == 100) {
+                builder.appendInt(valuesBlock.getInt(valueCount - 1));
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                var lowerValue = valuesBlock.getInt(lowerIndex);
+                var upperValue = valuesBlock.getInt(upperIndex);
+                var difference = (long) upperValue - lowerValue;
+                var percentileValue = lowerValue + (int) (fraction * difference);
+                builder.appendInt(percentileValue);
+            }
+        }
```

Comment:
Should this early return here? I might pull these out into a method that returns the value.

Reply:
Wow, good catch.
Some tests are failing in the CI because of weird things like pages having blocks with different position counts, so I think you found the culprit... 

And I thought I was testing sorted values in unit tests, but there may be something wrong there. Gonna check

Reply:
Added an assertion to AbstractScalarTestCase for the block positions, and it now failed successfully!